### PR TITLE
Cloudwatch: generate types

### PR DESCRIPTION
**Pull request by Clarity-89**
**What is this feature?**
Generate types for Cloudwatch data query. 

**Why do we need this feature?**

Part of the schematisation effort.

### CODE REVIEW

File: public/app/plugins/datasource/cloudwatch/composable_dataquery.cue [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+composableKinds: DataQuery: {
+	maturity: "merged"
+
+	lineage: {
+		seqs: [
+			{
+				schemas: [
+					{
+						common.DataQuery
+						//#CloudWatchMetricsQuery | #CloudWatchLogsQuery
```
Comment (Clarity-89) [inline public/app/plugins/datasource/cloudwatch/composable_dataquery.cue:34]: @sdboyer this doesn't seem to work, when I uncomment this I get an error `PluginTSTypesJenny: generating TS for child elements of schema failed: composableKinds.DataQuery.lineage.seqs.0.schemas.0: non-concrete value struct for input "cloudwatch"`.

File: public/app/plugins/datasource/cloudwatch/composable_dataquery.cue [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+						#CloudWatchMetricsQuery: {
```
Comment (Clarity-89) [inline public/app/plugins/datasource/cloudwatch/composable_dataquery.cue:35]: Just noticed that this actually extends a few other interfaces: `export interface CloudWatchMetricsQuery extends MetricStat, DataQuery`, how do we express that in cue?

File: public/app/plugins/datasource/cloudwatch/dataquery.cue [hunk 3]
```diff
@@ -0,0 +1,172 @@
+// Copyright 2023 Grafana Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License")
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
```
Comment (sarahzinger) [inline public/app/plugins/datasource/cloudwatch/dataquery.cue:7]: I'm not a licensing expert, but why would this file be under a different license than the rest of grafana?

**Review by sarahzinger — COMMENTED** [review]
Hi 👋🏻 this looks very neat, but I'm afraid I'm still a bit unclear how best to review these prs, do you have any advice? I see some things marked "todo doesn't work" is ok to merge it like that? How can I as a plugin maintainer help to validate this is a good implementation of a schema?

**Comment by Clarity-89** [comment]
> Hi 👋🏻 this looks very neat, but I'm afraid I'm still a bit unclear how best to review these prs, do you have any advice? I see some things marked "todo doesn't work" is ok to merge it like that? How can I as a plugin maintainer help to validate this is a good implementation of a schema?

Probably the best way to review this is to check if the generated types (TS types are in `dataquery.gen` file) look like the types they are replacing. The TODOs are for us to keep track of what needs to be fixed and for now will be manually fixed when we use the types. 

The schema added by this PR isn't used anywhere so it's not breaking anything. I'll try to actually use the generated types in a follow up PR, so then we might encounter some issues that need to be fixed or worked around. This PR is a setup for that so there's less code changes.

File: public/app/plugins/datasource/cloudwatch/dataquery.cue (hunk 3 above)
Comment (Clarity-89) [reply public/app/plugins/datasource/cloudwatch/dataquery.cue:7]: This is probably a question for @sdboyer.

**Review by sarahzinger — APPROVED** [review]
I think if this code is not being used yet, I do sort of wonder if it makes sense to add to the codebase, but on the otherhand I can see an argument for not wanting to make giant prs and breaking this work up into multiple steps!

Let us know if you'd rather the aws-plugins team look into this, or if you feel confident continuing to own this work. If you'd rather we take it over, it might make sense to make this a draft pr and we can create a design doc to do this work later in Q1, but if you'd rather take it on then I think we are very happy to merge this and support your efforts!

**Comment by Clarity-89** [comment]
> I think if this code is not being used yet, I do sort of wonder if it makes sense to add to the codebase, but on the otherhand I can see an argument for not wanting to make giant prs and breaking this work up into multiple steps!

Yeah, it's a matter of one huge PR vs several smaller ones. 

> Let us know if you'd rather the aws-plugins team look into this, or if you feel confident continuing to own this work. If you'd rather we take it over, it might make sense to make this a draft pr and we can create a design doc to do this work later in Q1, but if you'd rather take it on then I think we are very happy to merge this and support your efforts!

Ultimately the schema for each plugin is owned by the team that owns that plugin, grafana-as-code team provides the tooling needed to work with creating the schema and generating the types, but we don't have the domain expertise necessary to take ownership of the schemas, especially for complex plugins. A few plugin teams have already started converting their types to schemas and generating the types from them during the schema day. We used schema creation process for the Cloudwatch plugin to identify the issues with schema generation. I will try to continue this work in the follow up PR to see how these generated types work on practice but ultimately it'll be aws-plugins team that owns the schema definition here.

**Pull request merged**